### ISSUE TITLE
Creating relocatable binaries via AST

### ISSUE THREAD

**Issue by jayphelps** [WebAssembly/binaryen#1307]
It doesn't appear that binaryen's AST has a way to include [reloc sections](https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md#relocation-sections). What's everyone doing today with binaryen? (not using LLVM) Didn't see any discussions around a desire to add such support, so curious if perhaps I'm confused.

I guess this probably works in some limited situations as a temporary solution: `wasm2wat | wat2wasm --relocatable `  😆

**Comment by jayphelps** [comment]
To be clear, if it is something that makes sense in binaryen I would try to add it when I get the time. Not asking others to do the work for me 🕺

**Comment by kripken** [comment]
This should be added eventually, yeah, would be great if you want to do it. The relocations stuff is just getting stable now in LLVM, lld, etc., so it made sense to wait for that. But now is a good time.

We probably don't want to reimplement full linking in binaryen, but it would probably be useful to support loading/saving relocations. What use case did you have in mind?

**Comment by jayphelps** [comment]
@kripken Cool. use case: using binaryen as a backend to build wasm AST, emitting each unit to disk as relocatable binary, so we can then link them all together with any deps. Since this is pretty standard stuff I was worried that I'm missing something about binaryen since it seems like nearly anyone who it would want this. Want to make sure I'm not misunderstanding something.

**Comment by jgravelle-google** [comment]
For what it's worth I have some support for _reading_ reloc and linking sections in a tool I'm working on to generate emscripten-readable metadata (currently called lld-metadata, which will change before it lands) on the [o2wasm](https://github.com/WebAssembly/binaryen/tree/o2wasm) branch. If you want to use that in any way, it's here: https://github.com/WebAssembly/binaryen/blob/o2wasm/src/tools/lld-metadata.cpp

**Comment by dschuff** [comment]
@jayphelps I do think we want to support exactly the use case that you describe: allowing Binaryen to emit object files that are linkable with lld (and ideally, compatible with LLVM-generated binaries). That means writing relocations. I also agree with @kripken that we don't want to implement a linker; we should just use lld and I think it also should be able to handle the use cases that the current `wasm-merge` tool does. I also think we might want to support running wasm-to-wasm transforms and optimizations (i.e. the `opt` tool) on wasm object files; that would mean reading the relocations, and possibly making the optimizer aware of any conventions that would affect the ability to optimize.

**Comment by xujuntwt95329** [comment]
Hi, I'm wondering is there any progress on this direction?

The relocatable wasm binary is really helpful, if we want to build a compiler for advanced programming language, there must be a large amount of standard library or compiler-rt, it's common to write those libs using C/C++ and link with the user code to generate the final executable file. 

If binaryen lacks the feature to emit relocatable object files, then seems all the common libs need to be written using binaryen IR, this seems a huge amount of work to do.

**Comment by xujuntwt95329** [comment]
@kripken The Emscripten uses LLVM as codegen to generate the object files, and binaryen worked as an optimizer. 
However, we want to use binaryen directly as the codegen because we want to use the GC feature (currently seams GC opcodes not available in LLVM)

**Comment by kripken** [comment]
@xujuntwt95329 Ah, yes, so the issue is that LLVM doesn't support GC opcodes, so you can't run `wasm-ld` to link files that use GC? That is a current limitation.

Adding relocatable (wasm object file) support in Binaryen is one option to help there. But Binaryen would need to also be able to link the way `wasm-ld` does. That's a huge amount of work. Instead, I think a better option is to add a simple linker in Binaryen, `wasm-merge`. We had such a tool and removed it, but we could restore it. It would link files *without* relocation - what it does is connect imports to exports and combine two files into one. That's much simpler than `wasm-ld`, but I think it's enough, like this:

* Build C++ parts using LLVM. This emits object/relocatable files.
* Link C++ parts using `wasm-ld` to get a wasm `A`. This is a final wasm file.
* Build GC parts using Binaryen to get a wasm `B`. This is a final wasm file.
* Run `wasm-merge` to link `A` and `B` together. This is also a final wasm file.
* Run `wasm-opt` on that single final merged file (this will do inlining and other opts across C++ and GC code, DCE, etc.).

**Comment by xujuntwt95329** [comment]
Hi @kripken, thanks for your reply!

> Ah, yes, so the issue is that LLVM doesn't support GC opcodes, so you can't run wasm-ld to link files that use GC? That is a current limitation.

Yes, this is the current limitation.

And for the solution you proposed, yes, it is an interesting approach, it should work for some scenarios, but the limitation is the lack of relocation. 
- If we treat all value as GC object, don't use the linear memory, then seems relocation is not required; 
- However, if we want to use GC objects and linear memory together, then relocation is required (for global variables which are statically put to linear memory, also for the data section)

Currently the [WebAssembly Object File Linking](https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md) has already becomes a factual standard, and LLVM's `wasm-ld` fully support it, so IMO, if binaryen can generate these sections, we can benefit from the powerful `wasm-ld` tool, and then we can build more powerful tools based on binaryen.

**Comment by kripken** [comment]
Do your GC files actually need to touch linear memory directly? I wonder if they could call an API instead - an imported function that does a load from linear memory. That would become fast after linking + inlining. But you'd need to add similar workarounds for globals, etc.

If you do want to really mix relocations and GC in arbitrary ways then, yes, the only real solution is to either add GC support in `wasm-ld` or relocation support in Binaryen:
* I'm not sure how much GC support would be needed in `wasm-ld` - hopefully it would not need to actually read or understand the GC contents? cc @sbc100 
* Adding relocation support in Binaryen would take some work, but hopefully not much. We already use relocations for many things, like `call` has a name for the function and not an index, `global.get` has a name for the global, etc. What we are missing is stuff like relocations on immediates (like a `load` or `store` offset - we'd need to allow a name there and not just a constant) and on `const` nodes, etc.

**Comment by xujuntwt95329** [comment]
> Do your GC files actually need to touch linear memory directly?

Well, it is not decided yet, we are evaluating the capabilities of `binaryen` and `LLVM`, the functionality provided by the tools will largely influence our choice of how to use GC and linear memory. So this is also why I‘m asking, `relocation` is not a `must`, but if we support it, then there may be less limitation and more possibilities to use binaryen as a compiler codegen.

At this moment, I think we should start from the solution you mentioned before, since the relocation support seems will not ready in short time, but we are really looking forward to this feature as well as the DWARF debug information support, which is also a very important feature for compilers.

**Comment by kripken** [comment]
By "solution you mentioned before" do you mean the one using `wasm-merge`?

If so, let's wait to hear from @sbc100 on the above question about `wasm-ld`. If `wasm-ld` can link GC then we don't need `wasm-merge`. If it can't, then we do need `wasm-merge`, and then the question would be if there is someone with time&interest to work on it (it's not a lot of work, and I could get to it eventually, but I'm not sure when).

**Comment by tlively** [comment]
Maybe @ashleynh would be interested in working on wasm-merge due to its multi-memory component?

**Comment by sbc100** [comment]
`wasm-ld` has no support for GC today.  The work of adding such support would mostly be the introduction of new symbols types and new relocation types I believe... but my understanding on the GC proposal is fairly limited.

**Comment by sbc100** [comment]
I guess we would want to add support in the MC layer for writing wasm GC programs in assembly (if only so we can write wasm-ld tests), which I imagine is a good chunk of work too.

**Comment by xujuntwt95329** [comment]
> By "solution you mentioned before" do you mean the one using wasm-merge?

Yes, and before `wasm-merge` is ready, I think I can simply extract all the functions from one module through binaryen's existing API and insert them into the target module.

**Comment by tlively** [comment]
So far there has not been any demand from compilers using Binaryen as their backend to emit relocatable object files, so no one has worked on this. On the contrary, most new compilers seek to use binaryen as an optional component, which only works if Binaryen consumes and emits pure Wasm. I will close this as not planned for now, but of course we can reconsider in the future if a use case emerges.

**Comment by turboMaCk** [comment]
I'm implementing toy compiler where I'm experimenting in using bynarian as a backend. For now at least it doesn't even use binaryan IR just feeding wat to wasm-as. It's a toy in realy stages, nothing serious.

I was exploring the way I could link llvm code into the code I produce so naturally I came across this. `wat2wasm` provides `--relocatable` (as the issues description says) so it provides this feature. I think for my use-case something like wasm-merge would be enough. Reading this thread I think I have few options:

- simply not support linking to C like that
- build .o with wabt's wat2wasm and use binaryan only as an optional optimizer
- use wasm-merge
- implement my own linking (which would likely be like what wasm-merge seems to have been doing)

But as I say - it's nothing serious. I'm just sharing this so you have an idea about what use-case people might have. I'm putting this one on a shelf now - it's not something I need to decide right now.

**Comment by sbc100** [comment]
There has recently been more discussion of supporting relocatable object files in the wat format: https://github.com/WebAssembly/tool-conventions/pull/258

**Comment by sbc100** [comment]
I think we should also consider adding support for relocation information in inputs to binaryen so it can potentially optimize static memory segments.

The idea here would be that if/when binaryen removes dead code, it can also remove the corresponding dead data.   As of today binaryen has to tread data segments and essentially immutable.

I guess this is perhaps a different enough use case that it warrants a separate issue, but I'm not sure?
